### ISSUE TITLE
Feature Selection by FLAML?

### ISSUE THREAD

**Issue by knoam** [microsoft/FLAML#258]
Could you also use FLAML to select an optimal subset of features, perhaps using fewer features at first, then increasing, similar to how model complexity increases during training?

**Comment by jw00000** [comment]
From what I gleaned from playing with autosklearn, their approach was to build their search space to search for the best 'pipeline' where a pipeline included some preprocessing steps as well as the estimator. So included in the search space were hyperparameters defining choices about which preprocessing components to use and with what hyperparameters. Among the choices of preprocessing steps were many sklearn transformers including feature selection transformers. As a result the search space is rather large I think. I'm curious what you think about this approach. Would it make the search space too large to be practical? Do you think it would improve the quality of the models?

**Comment by qingyun-wu** [comment]
Hi @jw00000, thank you for sharing your experience with autosklearn and suggestions. Including the preprocessing component into the search space (as a hyperparameter with categorical choices) is a very reasonable approach, especially when the number of preprocessing choices is not that large, e.,g., it should be still practical when the number is less than 5.  Regarding the impact on model quality: if the time/resource budget is abundant, presumably the model quality won't become worse, although in the case where the time/resource budget is small, the quality of the resulting model may be degraded. We haven't tried this yet. Do you want to give it a try? We'd like to know how it works if so. Thank you!

**Comment by sonichi** [comment]
@knoam what is a metric you'd like to optimize when doing feature selection? With what constraint?

**Comment by jeet2005** [comment]
Coming back to this after 4 years quiet — I think the design @qingyun-wu
sketched in 2021 (feature count as a search-space hyperparameter, resolved
via a customized learner) is the right shape, and I want to propose a concrete
implementation now, because the piece that was missing in 2021 to do this
cleanly has since landed in the codebase. Posting the research below before
writing code, in the spirit of how #1200/#1564 were worked through here.

## What's changed since 2021 that makes this implementable now

@qingyun-wu's proposal was: add number-of-features as a hyperparameter, then
subset the data to that many features (by importance order) before the
estimator trains. The open problem in 2021 was *where* that subsetting step
belongs — `BaseEstimator` only had one preprocessing hook, mixing task-level
concerns (imputation, categorical encoding) with estimator-specific ones.

That's no longer true. The current `BaseEstimator` in `model.py` documents
and exposes preprocessing as two explicit levels:

```python
def preprocess(self, X):
    """
    This method applies estimator-specific preprocessing transformations...
    This is the second level of preprocessing that should be applied after
    task-level preprocessing (automl.preprocess()).
    """
    return self._preprocess(X)
```

This is exactly the hook a feature-selecting customized learner needs:
override `_preprocess` (and mirror it in `predict`) to subset columns to the
top-`k` by importance, *after* `DataTransformer` has already done imputation/
encoding, without touching task-level preprocessing at all. In 2021 this
would have meant patching shared preprocessing code; today it's a clean
subclass.

Second piece that's now available: every estimator (`sklearn`, `lightgbm`,
`xgboost`, `catboost`, linear models) exposes importances uniformly:

```python
@property
def feature_importances_(self):
    if hasattr(self._model, "feature_importances_"):
        return self._model.feature_importances_
    elif hasattr(self._model, "coef_"):
        return self._model.coef_
    else:
        return None
```

So a feature-selecting learner can rank by importance the same way regardless
of which underlying model it wraps — no per-estimator special-casing needed.

## Answering @sonichi's question ("what metric, what constraint?")

This was asked in 2021 and never resolved, and I think it's the actual reason
this stalled — without an answer, "add n_features as a hyperparameter" has no
concrete objective to optimize against. I'd propose: **no new metric at all**.
`n_features` becomes a search dimension inside the *existing* optimization
metric (accuracy/F1/log_loss/whatever the user already set) — CFO/BlendSearch
already treats early trials as cheap and expands them (see `low_cost_partial_config`
below), so a smaller feature count is just another axis of "cheap," resolved
by the existing search, not a new multi-objective problem.

For the "what constraint" half: FLAML already has `metric_constraints` in
`fit()` (`automl.py` line ~306) for exactly this shape of ask — e.g.
`pred_time_limit` is implemented as a metric constraint today. A user wanting
a hard cap (interpretability, latency, a fixed feature budget) could report
`n_features_used` from the custom metric function and add
`metric_constraints=[("n_features_used", "<=", 20)]`. No new mechanism needed
— this already exists and just needs a worked example in the docs.

## Answering @jw00000's autosklearn concern (search space blowup)

The autosklearn-style approach (categorical choice over multiple preprocessing
*pipelines*) is a different and much larger design than what's being proposed
here. A single `n_features` dimension added to one estimator's existing search
space (same shape as `subsample`/`colsample_bytree` already are for XGBoost)
adds one continuous/int dimension per estimator — not a combinatorial pipeline
choice. It stays inside FLAML's existing per-estimator search space size, so
I don't think the search-space-explosion concern applies to this version of
the design.

## Concrete proposal: reference `FeatureSelectingEstimator` mixin

Rather than changing every built-in estimator, ship one reference
implementation demonstrating the pattern (matching how customized learners
are already documented via `AutoML.add_learner`):

```python
class TopKFeatureSelectionMixin:
    """Mixin: adds a `n_features` hyperparameter that subsets columns
    to the top-k by importance from the *previous* trial of the same
    estimator type, before the wrapped estimator's own _fit/_preprocess."""

    @classmethod
    def search_space(cls, data_size, task, **params):
        base = super().search_space(data_size, task, **params)
        n_cols = data_size[1]
        base["n_features"] = {
            "domain": tune.lograndint(lower=max(1, n_cols // 10), upper=n_cols),
            "init_value": max(1, n_cols // 10),
            "low_cost_init_value": max(1, n_cols // 10),   # start cheap, grow
        }
        return base

    def _preprocess(self, X):
        X = super()._preprocess(X)
        if self._selected_columns is not None:
            X = X[self._selected_columns]
        return X

    def _fit(self, X_train, y_train, **kwargs):
        k = self.params.get("n_features", X_train.shape[1])
        if self._prior_importances is not None:
            self._selected_columns = self._rank_and_select(X_train.columns, k)
            X_train = X_train[self._selected_columns]
        train_time = super()._fit(X_train, y_train, **kwargs)
        self._prior_importances = self.feature_importances_
        return train_time
```

`low_cost_init_value` seeding `n_features` small directly implements the
literal request in this issue ("fewer features at first, then increasing"),
using the exact mechanism FLAML already uses for `n_estimators`/`num_leaves`
in `LGBMEstimator.search_space()` — so this isn't a new tuning concept, it's
reusing the existing cost-aware expansion machinery on a new axis.

Bootstrap question: first trial has no prior importances to rank by. Options:
(a) first trial uses all features, (b) first trial uses a cheap upfront rank
(e.g. mutual-info or a single fast model fit) shared across the estimator_list.
I'd lean (a) for simplicity in v1.

## Proposed benchmark before implementation

Given the precedent in #1200 (benchmark before shipping), I'd run:
- 2-3 datasets with genuinely redundant/noisy features (`make_classification`
  with `n_informative` << `n_features`, plus one real dataset with known-irrelevant
  columns).
- Compare: (A) no feature selection, (B) pre-applied `SelectKBest` workaround
  a user could do today, (C) this mixin.
- Metrics: validation score, wall-clock, *and* final feature count (since the
  benefit here is partly inference latency/interpretability, not just accuracy —
  which is also a reasonable answer to why this matters even when C ≈ B on
  accuracy alone).

## Open questions for maintainers
1. Does the `_preprocess`/`_fit` mixin shape above match how you'd want a
   reference implementation to look, or would you rather see this as a true
   built-in option on `LGBMEstimator` etc. directly rather than an opt-in
   mixin/customized learner?
2. Bootstrap strategy for the first trial — (a) or (b) above?
3. Worth documenting the `metric_constraints` route for hard feature-count
   caps as part of this, since it answers @sonichi's original question with
   something that already exists today?

Happy to run the benchmark and open a draft PR with the mixin once there's
alignment on the shape.
